### ISSUE TITLE
[wpe-2.28] Rewind after buffer removal doesn't succeed

### ISSUE THREAD

**Issue by varumugam123** [Metrological/WebKitForWayland#1064]
During testing of Xumo on Comcast STB, it is found that sometimes the rewind doesn't succeed and gets stuck in "[Seek] Delaying the seek: In async change PAUSED --> PLAYING"

Below is the sequence of operation done by application:
1) Playback is paused by app
2) Entire buffered range is removed, this triggers reenqueue() and eventually flush()
3) Flushing pipeline puts the pipeline in PAUSED->PAUSED async transition
4) App pushes data for seek position
5) App starts playback (i.e play()), this results in async PAUSED -> PLAYING transition
6) App initiates seek (this is ignored as there is an async transition in progress). The PLAYING transition doesn't complete as we don't have data for the current time

The issue could be reproduced with below test page as well: 
[test-app.txt](https://github.com/WebPlatformForEmbedded/WPEWebKit/files/11089793/test-app.txt)

Note : 
* Issue was observed with wpe-2.28 branch (373ad2332635bd5c56c679920e318d64ac6a9aef)
* Issue was not observed with wpe-2.22 as the flushing differs slightly there i.e both on buffer removal & append flush() is not performed

**Comment by calvaris** [comment]
FTR, from what I see in the logs (WebKit and console), steps 4 and 5 are swapped. Play is issued before appending the data.

Still investigating.

**Comment by calvaris** [comment]
@varumugam123 I just pushed https://github.com/WebPlatformForEmbedded/WPEWebKit/tree/calvaris/issue/1064 . Could you please try this and see if fixes the issue? I think it should.

**Comment by varumugam123** [comment]
Thanks for looking into this @calvaris, will check and let you know.

**Comment by calvaris** [comment]
@varumugam123 do you have any feedback yet?

**Comment by calvaris** [comment]
I'll end up close this if we don't get feedback.

**Comment by varumugam123** [comment]
Hi @calvaris Thank you for patience, please allow me to share the observations with this patch shortly

**Comment by varumugam123** [comment]
Hi @calvaris, below are my observations:
* I was not able to reproduce the issue on Broadcom devices, where the issue was originally reported, even without any workaround or this commit. I don't notice ASYNC PAUSED->PLAYING transition anymore with the test app (I guess something changed on the SOC recently)
* But I am still able to reproduce the issue on Realtek and applying your commit resolves it.

In summary, the patch fixes the issue reported.

**Comment by varumugam123** [comment]
Could you please help to clarify couple queries I have with the commit?
1) For buffered range the solutions works fine (i.e add data for the seeking position beforehand and perform seek), but if I rearrange the test app to seek then push buffer. I end up again in the indefinite wait.
2) m_isSwitchToPlayingOngoing will be true only when we request a state change to Playing and gets cleared when PLAYING state is achieved. Checking PAUSED_TO_PLAYING implies the same right?

**Comment by calvaris** [comment]
> Could you please help to clarify couple queries I have with the commit?
> 
>     1. For buffered range the solutions works fine (i.e add data for the seeking position beforehand and perform seek), but if I rearrange the test app to seek then push buffer. I end up again in the indefinite wait.

Can you please provide a test case for this?
 
>     2. m_isSwitchToPlayingOngoing will be true only when we request a state change to Playing and gets cleared when PLAYING state is achieved. Checking PAUSED_TO_PLAYING implies the same right?

From the pipeline PoV yes; from the player PoV I don't really now right now. I'll check it as soon as I get your modified test.

**Comment by varumugam123** [comment]
Hi @calvaris, I was not able to reproduce the case (1) I claimed anymore. I suppose something changed in the SoC. I believe we can mark this closed.
